consensus/core/src/transaction_certifier.rs

```diff
@@ -55,9 +67,40 @@ impl TransactionCertifier {
         }
     }
 
-    pub(crate) fn run_gc(&self) {
-        let gc_round = self.dag_state.read().gc_round();
+    /// Recovers internal state from blocks in storage.
+    /// Since votes are not stored persistently, blocks where votes have not been proposed are re-verified and voted.
+    pub(crate) fn recover(&self, block_verifier: &impl BlockVerifier) {
         let mut certifier_state = self.certifier_state.write();
+        let dag_state = self.dag_state.read();
+
+        let gc_round = dag_state.gc_round();
+        let authorities = certifier_state
+            .context
+            .committee
+            .authorities()
+            .map(|(index, _)| index)
+            .collect::<Vec<_>>();
+        for authority_index in authorities {
+            let blocks = dag_state.get_cached_blocks(authority_index, gc_round + 1);
+            let voted_blocks = blocks
+                .into_iter()
+                .map(|b| {
+                    if dag_state.is_hard_linked(&b.reference()) {
+                        (b, vec![])
+                    } else {
+                        let reject_transaction_votes = block_verifier
+                            .verify_and_vote(b.signed_block())
+                            .unwrap_or_else(|e| {
+                                panic!("Failed to verify block during recovery: {}", e)
+                            });
+                        (b, reject_transaction_votes)
+                    }
+                })
+                .collect::<Vec<_>>();
+            let certified_blocks = certifier_state.add_voted_blocks(voted_blocks);
+            self.send_certified_blocks(certified_blocks);
```

Comment:
This could actually end up sending duplicated certified blocks after recovery, correct? Not an issue, probably just have to dedupe on the Sui side.

i.e. Block A1 containing just Tx1 is rejected and certifed via fastpath but not yet proposed so it will get sent to Sui but after the node goes down and recovers it will resend this certified block to Sui?

Reply:
Yes similar to consensus commits, sending previously executed transactions is possible and ok.